Root specific lookup

carver opened the pull request:
### What was wrong?

Want to look up the data in a node without first switching the root hash, eg, update API from:

```py
original_root_hash = trie.root_hash
trie.root_hash = target_root_hash
assert trie.get(key) == expected_value_at_target_hash
trie.root_hash = original_root_hash
```

to:
```py
assert trie.get(key, target_root_hash) == expected_value_at_target_hash
```

### How was it fixed?

~~Builds on #83 -- this review can focus on just the last two commits (bf726c and e0c7ff5 currently) and I won't merge until #83 is approved.~~

Dropped a few places where `HexaryTrie` was relying on `self.root_hash` exclusively. Keep the old behavior for backwards compatibility.

Extended a current test for checking values against old roots.

#### Cute Animal Picture

![Cute animal picture](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSjqYYK1W_idKdlL4uqkxook4flCjdUDhaC1Hsn7bPCDyxN_b5w)

carver committed:
Enable root-specific lookups on trie

carver commented on tests/test_hexary_trie.py:
```diff
@@ -166,13 +166,20 @@ def unread_keys(self):
     for key, val in changes:
         if val is None:
             del trie[key]
+            missing_by_root[trie.root_hash].add(key)
```
Adding this value to a `set` instead of just assigning `= key` might be a little overboard, since the trie never goes back to an old state root right now. But it seemed easy enough to keep it general right now, in case we change the input data later.

carver commented on tests/test_hexary_trie.py:
```diff
@@ -316,3 +338,12 @@ def test_hexary_trie_missing_node():
+
+def test_hexary_trie_raises_on_pruning_snapshot():
+    db = {}
+    trie = HexaryTrie(db, prune=True)
```
Style flub due to copy-paste. Could combine previous lines since we don't need later access to `db`. I'll save that fix until after review.

lithp commented on trie/hexary.py:
```diff
@@ -535,6 +542,14 @@ def squash_changes(self):
 
+    @contextlib.contextmanager
+    def at_root(self, at_root_hash):
+        if self.is_pruning:
```
This seems a little dangerous, though I'm not sure how it could be fixed.

I guess the idea is: we're about to yield a copy of the trie and the two versions are likely to be used concurrently, if we allowed `is_pruning` to be true then operations such as `delete()` in one trie could mess with the other trie.

I scanned through `get()` and `set()` and some of the other methods and it seems this always works, `is_pruning` means the underlying trie won't change out from under us, but there's also nothing which enforces that being true. This is probably just me being paranoid, but "no operations on the other trie will change our nodes" and "we don't prune away outdated nodes" don't seem like equivalent concepts.

In particular, an optimization which modified nodes in place rather than building new ones and pruning the old ones would be allowed under the `is_pruning` interface but would break this method.

Definitely not suggesting this PR needs to do this! But: right now there are no restrictions on `self.db`, a solution might involve requiring `db` to be something which supports snapshotting. This would remove some potential for bugs and it'd also allow us to call `at_root` on tries which have pruning enabled.

lithp replied:
I'm sad because of Piper's suggestions I prefer the last two, the fluent interface is pretty!

The first one might be better though:

```
# local mutation
with trie.at_root(root_hash):
    trie.get(key)
```

If it mutates the old trie then there's no chance for users to accidentally concurrently modify a copy.

pipermerriam replied:
What about inverting this and having the copy do something like acquire a lock on the parent trie, and the parent trie refuses to prune if that lock isn't free?  That should allow more freedom of movement in this API while still providing roughly equal levels of safety.

carver replied:
> In particular, an optimization which modified nodes in place rather than building new ones and pruning the old ones would be allowed under the `is_pruning` interface but would break this method.

A trie that modifies in place (removing references to old nodes from the db) would certainly be "pruning" semantically, so `is_pruning` should be `True` there.

> Definitely not suggesting this PR needs to do this! But: right now there are no restrictions on self.db, a solution might involve requiring db to be something which supports snapshotting. This would remove some potential for bugs and it'd also allow us to call at_root on tries which have pruning enabled.

Yup, I am on board with doing this in future work. Having an immutable version of the trie has been on my wishlist for a long time. I listed it as a potential breaking change to make the default trie immutable: https://github.com/ethereum/py-trie/issues/85

> The first one might be better though:
> 
> ```
> # local mutation
> with trie.at_root(root_hash):
>     trie.get(key)
> ```
> If it mutates the old trie then there's no chance for users to accidentally concurrently modify a copy.

Here's my concern with that API. `root_hash` mutability can really mess with you in an async context:

```py
class TrieUser:
    def __init__(self, trie):
        self.trie = trie

    async def look_up_old_val(self, key, root):
        trie = self.trie
        with trie.at_root(root):

            # this checks if nodes are missing from the trie
            while busted(trie):
                await fix_up_trie(trie)  # this is adding nodes to the trie

            # this is returning the key at whatever root hash `trie` has
            return trie[key]

    def trie_update(self, key, val):
        # Updates might come in while awaiting above.
        self.trie[key] = val
```

If `trie_update()` runs while await-ing `fix_up_trie()`, then it:
1. modifies against a different state root than the class was initialized with
2. sets the new state root, which affects the `trie` await-ing inside the context above

lithp commented on trie/hexary.py:
```diff
@@ -8,7 +8,11 @@
 
-from eth_utils import to_list, to_tuple
+from eth_utils import (
+    ValidationError,
```
nitpick: `trie.exceptions` has a `ValidationError` which might be better

lithp commented on tests/test_hexary_trie.py:
```diff
@@ -1,19 +1,21 @@
 import json
 import os
 
 from eth_utils import (
```
nitpick of all nitpicks: I think our convention is to alphabetize these as well, maybe that could be part of this PR?

lithp commented on tests/test_hexary_trie.py:
```diff
@@ -166,13 +168,20 @@ def unread_keys(self):
 def test_hexary_trie_saves_each_root():
```
nitpick: This test now tests a couple different things.

Well, it seems like an oversight that it was previously only checking whether there were extra items at all! It's a nice addition that it also verifies that at least part of the older tries are accessible.

However `at_root` maybe deserves it's own test.

lithp reviewed (approved):
I have nothing but nitpicks and vague concerns, left you some feedback but feel free to :shipit: !

carver committed:
Review fixups

- Split big test in two
- Alphabetize some imports
- Use as_root in get_from_proof
- Use trie.exceptions.ValidationError instead of eth_utils

carver committed:
Add release notes
